### DISCUSSION TITLE
Expose cssnano configuration

### DISCUSSION THREAD

**Discussion by hatsumatsu** [vercel/next.js#22560] [Ideas]
**Describe the feature you'd like to request**

`cssnano` currently has a number of open issues that might not get fixed any time soon (e.g. https://github.com/vercel/next.js/issues/22389). These bugs affect production builds of next apps. Most of these bugs can be worked around by manually enabling / disabling cssnano optimizations. 

Unfortunately there does not seem to be a way to customize the cssnano config used by next and webpack.

**Describe the solution you'd like**

Exposing the config of `cssnano` via the config of webpack's  `CssMinimizerPlugin` in `next.config.js`
https://webpack.js.org/plugins/css-minimizer-webpack-plugin/#minimizeroptions

**Describe alternatives you've considered**

Using cssnano's config files could be another option:
https://cssnano.co/docs/config-file#cssnano-config-files

**Comment by hatsumatsu** [comment]
This is how I managed to configure webpack's CssMinimizerPlugin via `next.config.js`:

First we have to grab a copy of the plugin since `next` seems to use a built version with the config baked into it.

```
npm install css-minimizer-webpack-plugin@1.0.0 --save-dev
```
(Versions > 1.0.0 does not seem to work with next's bundled webpack version)

Add this to `next.config.js`:

```js
const CssMinimizerPlugin = require('css-minimizer-webpack-plugin')

module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {

    if( !dev ) {
      // remove existing instance of CssMinimizerPlugin
      config.optimization.minimizer = config.optimization.minimizer.filter( minimizer => {
        return minimizer.constructor.name !== 'CssMinimizerPlugin';
      });


      // create new instance of CssMinimizerPlugin
      // minimizerOptions === cssnano config
      const plugin = new CssMinimizerPlugin({
        minimizerOptions: {
          // your cssnano configuration
          preset: [
            'default',
            {
              mergeLonghand: false,
              mergeRules: false,
            },
          ],
        },
      });
      
      config.optimization.minimizer.push( plugin );
    }
   
    return config;
  }
}
```

It does work but probably breaks once next upgrades to a newer webpack version.

**Comment by lunelson** [comment]
I would really like to have this. cssnano's default transforms are very aggressive and they break certain modern/advanced CSS techniques. ~~The option of at least switching to the 'lite' preset would be helpful~~ Actually scratch that, even the 'lite' option is too naïve, it clobbers white-space custom properties for example, which are valid CSS

```css
/* this will be broken by cssnano, although empty space is a legitimate value */
:root { --my-prop: ; }
```
